File: retrofitrx/src/main/java/com/shuai/retrofitrx/config/NetConfig.java
package com.shuai.retrofitrx.config;

import android.app.Application;
import android.support.annotation.NonNull;

import com.shuai.retrofitrx.config.provider.NetBaseConfigProvider;
import com.shuai.retrofitrx.config.provider.NetRequestConfigProvider;
import com.shuai.retrofitrx.config.provider.NetResponseConfigProvider;
import com.shuai.retrofitrx.utils.Logger;


/**
 * 网络请求库的统一配置
 * <p>
 * api:
 * 获取Application：NetConfig.getApp();
 * 获取Config：NetConfig.getConfig.getXXXConfig().getXXX()
 */
public class NetConfig {
    private static NetConfig.Config sConfig;
    private static Application sApp;

    public static Config getConfig() {
        return sConfig;
    }

    public static Application getApp() {
        return sApp; // TODO 可能会有多线程问题。单例的解决。
    }

    /**
     * 初始化
     *
     * @param app Application
     * @return 配置文件
     */
    public static NetConfig.Config init(@NonNull final Application app) {
        sApp = app;
        if (sConfig == null) {
            sConfig = new NetConfig.Config();
        }
        return sConfig;
    }

    public static class Config {

        private NetBaseConfigProvider baseConfig = null;                    //基础配置：【必须配置】
        private NetRequestConfigProvider requestConfigProvider = null;      //请求配置：【必须配置】
        private NetResponseConfigProvider responseConfigProvider = null;    //响应配置：可选配置

        public Config baseConfig(NetBaseConfigProvider baseConfig) {
            this.baseConfig = baseConfig;
            Logger.debug(baseConfig.isLogDebug());//设置log日志模式
            return this;
        }

        public Config requestConfig(NetRequestConfigProvider requestConfigProvider) {
            this.requestConfigProvider = requestConfigProvider;
            return this;
        }

        public Config responseConfig(NetResponseConfigProvider responseConfigProvider) {
            this.responseConfigProvider = responseConfigProvider;
            return this;
        }


        //////////////////////////////////GET AND SET////////////////////////////////////////

        public NetBaseConfigProvider getBaseConfig() {
            return baseConfig;
        }

        public NetRequestConfigProvider getRequestConfigProvider() {
            return requestConfigProvider;
        }

        public NetResponseConfigProvider getResponseConfigProvider() {
            return responseConfigProvider;
        }

        //////////////////////////////////GET AND SET////////////////////////////////////////


    }
}

File: retrofitrx/src/main/java/com/shuai/retrofitrx/config/NetConfig.kt
package com.shuai.retrofitrx.config

import android.app.Application
import com.shuai.retrofitrx.config.provider.NetBaseConfigProvider
import com.shuai.retrofitrx.config.provider.NetRequestConfigProvider
import com.shuai.retrofitrx.utils.Logger

/**
 * 网络请求库的统一配置
 *
 *
 * api:
 * 获取Application：NetConfig.getApp();
 * 获取Config：NetConfig.getConfig.getXXXConfig().getXXX()
 */
class NetConfig {

    companion object {

        @JvmStatic
        val config: Config by lazy {
            Config()
        }


        @JvmStatic
        lateinit var app: Application
            private set

        /**
         * 初始化
         *
         * @param app Application
         * @return 配置文件
         */
        @JvmStatic
        fun init(app: Application): Config {
            NetConfig.app = app
            return config
        }

        class Config {
            //基础配置：【必须配置】
            lateinit var baseConfigProvider: NetBaseConfigProvider
                private set

            //请求配置：【选择配置，也可以在单独请求里传入配置】
            var requestConfigProvider: NetRequestConfigProvider? = null
                private set

            fun baseConfig(baseConfigProvider: NetBaseConfigProvider): Config {
                this.baseConfigProvider = baseConfigProvider
                Logger.debug(baseConfigProvider.isLogDebug) //设置log日志模式
                return this
            }

            fun requestConfig(requestConfigProvider: NetRequestConfigProvider?): Config {
                this.requestConfigProvider = requestConfigProvider
                return this
            }
        }
    }

}